texlive: multi-output packages (subset of #250805)

xworld21 opened the pull request:
## Description of changes
For ease of reviewing and merging (or god forbid, reverting), this is the first half of #250805: rewrite TeX Live packages as pseudo-multi-output derivations.

This is fully backward compatible on its own as it works with the original `texlive.combine`, and it causes no rebuild.

I believe this is as complete (it fools `buildEnv`, and splicing works) and as backward compatible as it can be, with little freedom in terms of design space. I believe the only non-standard interface is `tlOutputName`, because fixed output derivations do *not* support changing the output name, which is moderately annoying.

## Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

apfelkuchen6 commented on pkgs/tools/typesetting/tex/texlive/build-texlive-package.nix:
```diff
@@ -15,126 +15,192 @@
+  // lib.optionalAttrs hasSource { texsource = texsource; }
+  // lib.optionalAttrs hasTlpkg { tlpkg = tlpkg; }
+  // lib.optionalAttrs hasManpages { man = man; }
+  // lib.optionalAttrs hasInfo { info = info; };
```
I'd love if hydra would build all the containers automatically once this is exposed as `texlivePackages`, so we don't have problems like missing docs on the binary cache.

If the derivation contains only the bin container (or the run container if the bin container doesn't exist), and all the other containers are just passthru attributes, hydra will only touch the primary. containers. Do you have any idea how this could be achieved? It's not possible to have actual multi-output-derivations where just some outputs have a fixed hash, right?

apfelkuchen6 commented on pkgs/tools/typesetting/tex/texlive/generate-fixed-hashes.nix:
```diff
@@ -5,7 +5,7 @@ with lib; let
   # ugly hack to extract combine from collection-latexextra, since it is masked by texlive.combine
   combine = lib.findFirst (p: (lib.head p.pkgs).pname == "combine") { pkgs = [ ]; } (lib.head texlive.collection-latexextra.pkgs).tlDeps;
-  all = filter (p: p ? pkgs) (attrValues (removeAttrs texlive [ "bin" "combine" "combined" "tlpdb" ])) ++ [ combine ];
+  all = filter (p: p ? pkgs) (attrValues (removeAttrs texlive [ "bin" "combine" "combined" "tlpdb" ])) ++ [ { inherit (combine) pkgs; } ];
```
Might as well get rid of that abomination now and use the internal interface `__pkgs`, this makes my eyes bleed.

xworld21 replied:
done!

apfelkuchen6 commented on pkgs/tools/typesetting/tex/texlive/build-texlive-package.nix:
```diff
@@ -15,126 +15,192 @@
+  };
+
+  # TODO switch to simpler "${name}" (requires a rebuild)
+  out = runCommand "texlive-${pname}.bin-${version}"
```
```suggestion
  bin = runCommand "texlive-${pname}.bin-${version}"
```

I think `bin` makes it a bit easier to understand what's going on. (also in the last line)

xworld21 replied:
I switched back and forth on this... changed it back to `bin`. Still `out` for the outside world of course.

apfelkuchen6 reviewed (approved):
Looks really good!

xworld21 commented:
![image](https://github.com/NixOS/nixpkgs/assets/1962985/56770c8e-efd1-42a0-acb8-f7969d62c1b0)
At last! And we'll get more with `buildEnv` and the setup hooks.

apfelkuchen6 commented:
Hm, I cannot reproduce this eval performance improvement. In fact, we appear to have a recent regression:

Instantiatiating `texlive.combined.scheme-full`  takes:
this PR:  2.1s
current master: 2.0s
after [bin containers](https://github.com/NixOS/nixpkgs/commit/aa725d5a5bc3e782cf7cb872616c9c3cf5330162): 1.8s
before bin containers, 1.5s

Bisect between the bin containers and master is running.

apfelkuchen6 commented:
Ok, the eval performance regression is due to recent python changes. Nothing we can do about that:

only skipped commits left to test
possible first bad commit: [e1cafe642a5e7d265d7b1dc55e2ddda81d219377] python3.pkgs.build: move tests to passthru
possible first bad commit: [349f57fb426225483f12e082e5ab5da8137eefe8] python3.pkgs.pyproject-hooks: move tests to passthru
possible first bad commit: [0d7647889c6701d8a0e6fa784e886012011ca2d8] python3.pkgs.installer: move tests to passthru
possible first bad commit: [6c85fff302615c62bf4f632bca661bc48298b0a3] python3.pkgs.buildPythonPackage: switch to PyPA build and installer

apfelkuchen6 commented on pkgs/tools/typesetting/tex/texlive/build-texlive-package.nix:
```diff
@@ -15,17 +15,27 @@
+
+# TODO stabilise a generic interface decoupled from the finer details of the
+# translation from texlive.tlpdb to tlpdb.nix
+lib.makeOverridable({ pname
```
We might want to replace `lib.makeOverridable` with a variant of it that automatically unlocks the fixed hashes.
 
This doesn't have to be in this PR, any time before stabilizing (and announcing) the interface will do, but I wanted ot mention it somewhere.

On that note, we probably want individual `postUnpack`/`postFixup`-thingies for each of the containers (or possibly something that takes an attrset).

xworld21 replied:
Good points. Re fixed hashes, we have to plan it carefully: some operations are hash safe and invalidating the hash would be surprising (like changing `postFixup` or the formats).

xworld21 replied:
For clarity: I removed `makeOverridable` because the interface of `buildTeXLivePackage` is still unstable, so I am marking this as resolved. It will need to be addressed, but presumably after 23.11.

apfelkuchen6 commented on pkgs/tools/typesetting/tex/texlive/default.nix:
```diff
@@ -100,12 +100,41 @@ let
+
+  # combine a set of TL packages into a single TL meta-package
+  combinePkgs = pkgList: lib.catAttrs "pkg" (
+    let
+      # a TeX package is an attribute set { pkgs = [ ... ]; ... } where pkgs is a list of derivations
+      # the derivations make up the TeX package and optionally (for backward compatibility) its dependencies
+      tlPkgToSets = drv: map ({ tlType, version ? "", outputName ? "", ... }@pkg: {
+          # outputName required to distinguish among bin.core-big outputs
+          key = "${pkg.pname or pkg.name}.${tlType}-${version}-${outputName}";
+          inherit pkg;
+        }) (drv.pkgs or (toTLPkgList drv));
+      pkgListToSets = lib.concatMap tlPkgToSets; in
+    builtins.genericClosure {
+      startSet = pkgListToSets pkgList;
+      operator = { pkg, ... }: pkgListToSets (pkg.tlDeps or []);
+    });
```
This is now (after #250621) unused.

xworld21 replied:
Ahem, removed.

xworld21 committed:
texlive.buildTeXLivePackage: switch to fake multi-output derivations for TeX Live packages

xworld21 committed:
tests.texlive.binaries: use new texlive package source

xworld21 committed:
texlive.combine: only derivations under .packages

xworld21 committed:
texlive.tlpdb.nix: detect presence of info pages

xworld21 committed:
mftrace: use new tex package structure

xworld21 committed:
biber-ms: use new texlive package source

xworld21 committed:
texlive.bin.xpdfopen: use new texlive package source

xworld21 committed:
biber: use new texlive package source

xworld21 committed:
texlive.bin.pygmentex: use new texlive package source

xworld21 committed:
sagetex: use new tex package structure

xworld21 committed:
tests.texlive.shebangs: use new texlive package source

xworld21 committed:
texlive: document new texlive.pkgs attribute

xworld21 committed:
iwona: use new texlive package source

xworld21 commented:
Rebased against new `texlive.bin.core` changes. I decided to flip the switch: there is no scenario without `texlive.pkgs`, so I made that official in the docs and applied the relevant in the rest of the tree. I also added a fix to make `texlive.combine` vaguely aware of `outputSpecified` – but that would be a new usage, so it will most likely be superseded by `withPackages`.

(I am also curious to see the *actual* evaluation report against the latest master.)

xworld21 commented on pkgs/misc/sagetex/default.nix:
```diff
@@ -15,8 +16,14 @@ stdenv.mkDerivation (finalAttrs: rec {
+
+  nativeBuildInputs = [
     texlive.combined.scheme-basic
+    # multiple-outputs.sh fails if $out is not defined
+    (writeShellScript "force-tex-output.sh" ''
+      out="''${tex-}"
+    '')
```
This is the ugly bit: it's tricky to generate a TeX package with only output `tex`. We should offer a generic hook `texlive.forceTeXOutput`... ?
